### PR TITLE
Yufei - make intangible time numbers having a mouseover explanation editable by the owner

### PR DESCRIPTION
**Pull request by yufeiStella**
# Description
Screen Shot 2023-07-21 at 16 31 39

## Related PRS (if any):
This frontend PR is related to the [#454](https://github.com/OneCommunityGlobal/HGNRest/pull/454) backend PR.

## Main changes explained:
- Create mouseover text of total time component and add it into leaderboard component
- Create mouseover text action, reducer, etc.

## How to test:
1. check into current branch
2. do `npm install` and `npm run start:local` to run this PR locally
3. log in as an owner user and verify that the mouseover text appears when you move your mouse over both 'Total Time' and the numbers on the Leaderboard. Some numbers are blue (indicating intangible hours logged), and some numbers are black. Mouseover text will appear on both types of numbers.
4. Only the owner can see the edit button next to 'Total Time'. Click on it and enter the new text to update the mouseover text. Please note that it may take some time to respond, as shown in my screen recording.
5. Wait a few seconds for a green pop-up box to appear, showing the message "Mouseover Text updated!" Then you will see the newly updated mouseover text when you move your mouse over both 'Total Time' and the numbers. If you don't immediately see the changes, just wait a few more seconds, as it may take some time for the updates to respond.
6. Hard refresh the page to see the change is globally. The same thing is if you don't immediately see the changes, just wait a few more seconds, as it may take some time for the updates to respond.
7. login as any other role to verify the mouseoverText updates globally and they do not see the edit button.

## Screenshots or videos of changes:
Log in as an owner:
Screen Shot 2023-07-21 at 19 43 12

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111897624/4ceff3aa-ae86-4ba2-8dfa-096808022cb5

Log in as any other user except owner:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111897624/46e3df42-b839-4469-bd34-afc8113ecf54

## Note:
Sometimes, you may need to wait for the page to respond and display the mouseover text. I encountered a similar situation as shown in my screenshot recording. When you move your mouse to another place where there should be mouseover text, you may experience the same situation. Just be patient and wait.

### CODE REVIEW

**Review by xaanders — CHANGES_REQUESTED** [review]
Hi @yufeiStella !
I have tested your PR. Functionality works as intended. 
I could change mouseover text as an owner
 



As an admin or manager I could only read the explanation: 



However I want to suggest rendering of the button based on permissions(and probably sending the update request as well):


[scrnli_7_22_2023_10-41-35 AM.webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/108314277/f9f5a724-11b0-4b5e-b1da-e6d9ced775b2) 


change

**Comment by igorrochadasilva** [comment]
Hi @yufeiStella !
I hope you are well =D

I tested the PR, the admin I can see the mouseoverText, and can't see the edit button.

1030_01_admin-mouseover

But when I tested logged in as Owner, I can't see the edit button, I'll try to see what happened, maybe something in my local environment
1030_02_owner-mouseover

**Review by n0vay — COMMENTED** [review]
Great Job!
Checked the file as an owner to edit the hover over text, was able to see it with admin and other roles account too.
One small suggestion, you should include the steps to change backend in the testing instructions, I was a bit confused when it was not working.
image

**Review by AuroraLovesWhales — DISMISSED** [review]
Hi @yufeiStella. I have tested your PR and it is working as expected. Great job!

Log in as Owner, edit the mouseover text for total time, and refresh the page:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44281916/0fd26ffd-c936-421e-8a17-701d65178784

Log in as any users other than Owner:
image

**Review by AaronPersaud — CHANGES_REQUESTED** [review]
Hi, I've tested it and it works as expected!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/10265513/fec3a984-ed20-4ffa-81ed-9a87d893e8af

(I realized after the video didn't record cursor hovering and displaying the text)

I tested the following:
- non owners cannot see icon for editing
- both textareas show up for all users
- owner can edit and save text
- text is updated on refresh and when switching to another page and back

I have added some code suggestions that would make it cleaner and a bit more concise.

**Review by shiva2096 — APPROVED** [review]
Hi @yufeiStella 

Great work! Functionality is working as expected.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/26687177/84c09e2a-a593-4815-803e-522fdaef273a

**Review by winghojackyli — APPROVED** [review]
I tested your PR and it worked as expected. 
I logged in as owner and I can change the mousehover text from "text" to "text by owner". 

Screenshot (1)

Screenshot (2)

Screenshot (3)

Then, I logged in as admin and I do not have the edit mousehover text button, but I can see the edited mousehover text. 

Screenshot (4)

I logged in as volunteer and it worked as expected as well. 

Screenshot (5)

**Comment by AnishPandita** [comment]
Hi @yufeiStella it seems to work correctly. Edited using Owner account and verified by admin account as well.
Screenshot 2023-07-24 at 5 25 36 PM
Screenshot 2023-07-24 at 5 31 50 PM
Screenshot 2023-07-24 at 5 33 07 PM

**Review by DavidC0126 — DISMISSED** [review]
Hi @yufeiStella, I have tested your PR and everything works as intended! 
I am able to see and edit the mouseover text as as owner, and I can only see the edited text but not able to edit the test as any other role.

-  Owner perspective - edit icon and ability to adit the text
PR1030_owner

PR1030_editedWords

- Admin role perspective => can see but cannot edit it
PR1030_adminRole

**Comment by beblicarl** [comment]
I tried several times but I was unsuccessful in getting this to work as intended. Also, there were errors in the backend console which could be related to this issue.

You would notice from the video that even after editing the field with different texts the text that shows onMouseOver remains "some text:text"

[screencast-localhost_3000-2023.07.24-17_04_18.webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/68056104/67704d45-fa41-41c6-9098-5fe5337334c7)

issue #1030-error

**Review by eduvarjaor — APPROVED** [review]
Disconsidering the slow local host, this PR works as intended. Great job!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/86692035/399b6de8-0950-4758-a2e0-bfc76608cfb6

**Review by sxiong5 — DISMISSED** [review]
Hey, the PR worked as intended.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/62367016/c41ffbce-94c7-4063-b73a-7085f4c33241

**Comment by Shreyj14** [comment]
Hi @yufeiStella

Reviewed this PR with backend 454, and it works as intended.
Screenshot 2023-07-26 at 6 10 53 PM

Screenshot 2023-07-26 at 6 11 55 PM
Screenshot 2023-07-26 at 6 12 04 PM
Screenshot 2023-07-26 at 6 13 00 PM

**Review by Shreyj14 — APPROVED** [review]

**Comment by yufeiStella** [comment]
> Hi @yufeiStella ! I have tested your PR. Functionality works as intended. I could change mouseover text as an owner
> 
>  
> 
> As an admin or manager I could only read the explanation: 
> 
> However I want to suggest rendering of the button based on permissions(and probably sending the update request as well):
> 
>  scrnli_7_22_2023_10-41-35.AM.webm 
> change

Hi @AaronPersaud, thank you for your suggestion! I wonder what means rendering of the button based on permissions? The visibility is similar with the owner message functionality located in the bar of the dashboard, only owner will see the edit button, so no need to worry changing the permission within admin, manager, etc.

**Comment by yufeiStella** [comment]
> 

Hi @beblicarl, I'm not sure if the error is related to my functionality, but it does have a long response time or even have to reload it again when running the app. Besides, I noticed that you don't have any warnings after running 'npm start' and so you can receive the feedback in the backend console. However, every time I run 'npm start', I receive the warning (Use `node --trace-deprecation ...` to show where the warning was created" after running 'npm run build' and 'npm start' in the backend. The application can still run, but then I can't see any feedback in the backend console when running the application on localhost. Have you ever had the same warning (because I know someone else also have the same problem), and how did you resolve it? Or have you never encountered such a problem before?

**Comment by yufeiStella** [comment]
> I tried several times but I was unsuccessful in getting this to work as intended. Also, there were errors in the backend console which could be related to this issue.
> 
> You would notice from the video that even after editing the field with different texts the text that shows onMouseOver remains "some text:text"
> 
>  screencast-localhost_3000-2023.07.24-17_04_18.webm 
> issue #1030-error


Hi @beblicarl, I'm not sure if the error is related to my functionality, but it does have a long response time or even have to reload it again when running the app. Besides, I noticed that you don't have any warnings after running 'npm start' and so you can receive the feedback in the backend console. However, every time I run 'npm start', I receive the warning (Use node --trace-deprecation ... to show where the warning was created" after running 'npm run build' and 'npm start' in the backend. The application can still run, but then I can't see any feedback in the backend console when running the application on localhost. Have you ever had the same warning (because I know someone else also have the same problem), and how did you resolve it? Or have you never encountered such a problem before?

**Comment by iTvX** [comment]
I have checked your PR. The owner can change the text, and it looks good. But when I switch accounts (not the owner's account), I cannot see the text. It seems like it's not global.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/46791134/cd5d1198-0e63-4f84-a35d-92400f833841

**Review by yurisokolovicz — APPROVED** [review]
Hi @yufeiStella,

Your new functionality is working as intended, but we are experiencing an infinite scroll problem in the front-end. Sometimes, it is very difficult to see the functionality due to this issue. However, please note that this problem is not related to your work and it will be fixed this week. I approve the PR. Good job!

Screenshot from 2023-08-01 10-47-01

File: src/components/LeaderBoard/Leaderboard.jsx [hunk 1]
```diff
@@ -204,9 +212,21 @@ const LeaderBoard = ({
...
+              <th style={{ textAlign: 'right' }}>
+                <div style={{ display: 'flex', alignItems: 'center' }}>
+                  <div style={{ textAlign: 'left' }}>
+                    <span className="d-sm-none">Tot. Time</span>
+                    <span className="d-none d-sm-inline-block" title={mouseoverText}>Total Time </span>
+                  </div>
+                  {isOwner && (
+                    <MouseoverTextTotalTime onUpdate={handleMouseoverTextUpdate} />
+                  )}
+                  {!isOwner && (
```
[the remark this answers is not in the document]
Comment (yufeiStella) [reply src/components/LeaderBoard/Leaderboard.jsx:225]: Thank you! Updated my code. The thing is that I still need to fetch updated mouseover text, but now I have changed to the following code:  
{isOwner && (
                    <MouseoverTextTotalTimeButton onUpdate={handleMouseoverTextUpdate} />
                  )}
                  {!isOwner && (
                    <MouseoverTextTotalTimeUpdater onUpdate={handleMouseoverTextUpdate} />
                  )}
MouseoverTextTotalTimeUpdater is just to fetch updated data, not a button with update function.

[the remark this answers is not in the document]
Comment (yufeiStella) [reply src/components/LeaderBoard/Leaderboard.jsx:225]: All set, thank you for suggestion!

File: src/components/mouseoverText/mouseoverTextTotalTime.jsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            setModalOpen(false);
+            setText(newText);
+            onUpdate(newText);
+        } else {
+            //console.log('createMouseoverText is ' + mouseoverText.newMouseoverText);
+            createMouseoverText(mouseoverText);
+            setModalOpen(false);
+            setText(newText);
+            onUpdate(newText);
+        }
```
[the remark this answers is not in the document]
Comment (yufeiStella) [reply src/components/mouseoverText/mouseoverTextTotalTime.jsx]: All set, thank you for suggestion!

[the remark this answers is not in the document]
Comment (yufeiStella) [reply src/components/mouseoverText/mouseoverTextTotalTime.jsx]: I've just updated this PR. Could you please re-review my PR and approve it if you agree with the changes? Thank you!

**Comment by yufeiStella** [comment]
> Hi @yufeiStella ! I have tested your PR. Functionality works as intended. I could change mouseover text as an owner
> 
>  
> 
> As an admin or manager I could only read the explanation: 
> 
> However I want to suggest rendering of the button based on permissions(and probably sending the update request as well):
> 
>  scrnli_7_22_2023_10-41-35.AM.webm 
> change

Hi @xaanders, now I have rendered the button based on permissions(also the updating request because it is all about the edit button)! Could you please re-review it again and approve it if you agree with it? Thank you!



https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111897624/4b0f0d40-abe0-4ea1-86f4-90dc9258c620

**Comment by beblicarl** [comment]
The functionality works as intended and the code is great!

[screencast-bpconcjcammlapcogcnnelfmaeghhagj-2023.08.02-18_28_12.webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/68056104/cef6846b-eb09-4e08-b5f7-126c3422aadb)

**Review by beblicarl — DISMISSED** [review]

File: src/components/LeaderBoard/Leaderboard.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (AaronPersaud) [reply src/components/LeaderBoard/Leaderboard.jsx:225]: Thank you, looks good.

**Review by AaronPersaud — APPROVED** [review]
Hi Yufei, I have tested it and it looks good!

Thank you for addressing my code suggestions, they look great.

Nice work!

File: src/components/LeaderBoard/Leaderboard.jsx [hunk 3]
```diff
@@ -30,15 +31,32 @@ function useDeepEffect(effectFunc, deps) {
...
   const isAdmin = ['Owner', 'Administrator', 'Core Team'].includes(loggedInUser.role);
+  const isOwner = ['Owner'].includes(loggedInUser.role);
+
+  const [mouseoverTextValue, setMouseoverTextValue] = useState(totalTimeMouseoverText);
+
+  useEffect(() => {
+    getMouseoverText();
+  }, [totalTimeMouseoverText]);
+
+  useEffect(() => {
```
Comment (xaanders) [inline src/components/LeaderBoard/Leaderboard.jsx:53]: I would suggest to make just one useEffect because they have the same dependency anyway

Comment (yufeiStella) [reply src/components/LeaderBoard/Leaderboard.jsx]: Thank you! Fixed.

**Review by xaanders — APPROVED** [review]

**Pull request merged**